Kernel/KBO.hpp

```diff
@@ -249,24 +260,13 @@ class KBO
 
     friend class KBOComparator;
 
-    int _weightDiff;
-    /** The variable counters */
-    DHMap<unsigned, int, IdentityHash, DefaultHash> _varDiffs;
-    /** Number of variables, that occur more times in the first literal */
-    int _posNum;
-    /** Number of variables, that occur more times in the second literal */
-    int _negNum;
-    /** First comparison result */
-    Result _lexResult;
-    /** The ordering used */
-    KBO& _kbo;
   }; // class KBO::State
 
   /**
    * State used for comparing terms and literals
    */
-  mutable State* _state;
+  mutable std::unique_ptr<State> _state;
```

Comment:
Is there a reason this is a pointer at all? Looks like it could just be a field.

Reply:
Good question .. I don't really know whay it was a pointer in the first place .. I just kept it as it was I think to not mess up any magic :)

Reply:
I know there is some part of the code base that sets the pointer to `NULL` in order to be able to `ASSERT` that it's not being written to by multiple pocedures at the same time. But that's only debugging code. So I'm not sure..  maybe @mezpusz knows why it is a pointer at all?

Reply:
I don't know where the code originates from or what the intention was. If thread safety was the issue it would be better to have the `State`s recycled in my opinion, otherwise setting them to null does not make too much sense. What do you think?

Reply:
OK - let's keep it as-is for now to preserve the magic, but remember that it's weird and slate it for making it ...maybe just a `static` or plain members at some future point?